test/e2e/framework/resource_creator.go

```diff
@@ -104,80 +107,38 @@ func (ctx *TestCtx) createCRFromYAML(yamlFile []byte, resourceName string) error
 	return err
 }
 
-func (ctx *TestCtx) createCRDFromYAML(yamlFile []byte) error {
-	decode := extensions_scheme.Codecs.UniversalDeserializer().Decode
-	obj, _, err := decode(yamlFile, nil, nil)
+func setNamespaceYAML(yamlFile []byte, namespace string) ([]byte, error) {
+	yamlMap := make(map[interface{}]interface{})
+	err := yaml.Unmarshal(yamlFile, &yamlMap)
 	if err != nil {
-		return err
-	}
-	switch o := obj.(type) {
-	case *crd.CustomResourceDefinition:
-		_, err = Global.ExtensionsClient.ApiextensionsV1beta1().CustomResourceDefinitions().Create(o)
-		ctx.AddFinalizerFn(func() error {
-			err = Global.ExtensionsClient.ApiextensionsV1beta1().CustomResourceDefinitions().Delete(o.Name, metav1.NewDeleteOptions(0))
-			if err != nil && !apierrors.IsNotFound(err) {
-				return err
-			}
-			return nil
-		})
-		if apierrors.IsAlreadyExists(err) {
-			return nil
-		}
-		return err
-	default:
-		return errors.New("non-CRD resource in createCRDFromYAML function")
+		return nil, err
 	}
+	yamlMap["metadata"].(map[interface{}]interface{})["namespace"] = namespace
+	return yaml.Marshal(yamlMap)
 }
 
 func (ctx *TestCtx) CreateFromYAML(yamlFile []byte) error {
-	yamlMap := make(map[interface{}]interface{})
-	err := yaml.Unmarshal(yamlFile, &yamlMap)
+	namespace, err := ctx.GetNamespace()
 	if err != nil {
 		return err
 	}
-	kind := yamlMap["kind"].(string)
-	switch kind {
-	case "Role":
-	case "RoleBinding":
-	case "Deployment":
-	case "CustomResourceDefinition":
-		return ctx.createCRDFromYAML(yamlFile)
-	// we assume that all custom resources are from operator-sdk and thus follow
-	// a common naming convention
-	default:
-		return ctx.createCRFromYAML(yamlFile, strings.ToLower(kind)+"s")
-	}
+	yamlSplit := bytes.Split(yamlFile, []byte("\n---\n"))
+	for _, yamlSpec := range yamlSplit {
+		yamlSpec, err = setNamespaceYAML(yamlSpec, namespace)
+		if err != nil {
+			return err
+		}
 
-	decode := scheme.Codecs.UniversalDeserializer().Decode
-	obj, _, err := decode(yamlFile, nil, nil)
-	if err != nil {
-		return err
-	}
+		obj, _, err := Global.DynamicDecoder.Decode(yamlSpec, nil, nil)
+		if err != nil {
+			return ctx.createCRFromYAML(yamlFile)
```

Comment:
why are we calling `createCRFromYAML` if decoding has failed?

Reply:
In the e2e framework, the dynamic client (and thus the dynamic decoder) can handle everything that kubeclient and extensions client can. If decoding fails, that means it is not a standard kubernetes resource or a resource in extensions. This means that it is most likely a custom resource, so we use handle it with createCRFromYAML. The external test framework does not need this since the api for the custom resources is available at compile time so we can just use AddToFrameworkScheme. However, since we generate a test project during the internal e2e tests, we do not have the api available to use at compile time during these tests.

Essentially, we are using createCRFromYAML as a fallback for unrecognized resources, which is what we did before this PR as well.

Reply:
I see. that makes sense. Could you add a comment?